### ISSUE TITLE
Add integration support for energidataservice to get prices

### ISSUE THREAD

**Issue by cnrd** [jonasbkarlsson/ev_smart_charging#87]
Add support for getting prices from https://github.com/MTrab/energidataservice

The format is very like nordpool, but differs in using `price` instead of `value` and using `hour` which is equal to `start` (`end` is implicit)

Format of raw_today:

```yaml
raw_today: 
- hour: '2023-01-02T00:00:00+01:00'
  price: 0.92
- hour: '2023-01-02T01:00:00+01:00'
  price: 0.862
- hour: '2023-01-02T02:00:00+01:00'
  price: 0.873
- hour: '2023-01-02T03:00:00+01:00'
  price: 0.792
- hour: '2023-01-02T04:00:00+01:00'
  price: 0.847
- hour: '2023-01-02T05:00:00+01:00'
  price: 1.039
- hour: '2023-01-02T06:00:00+01:00'
  price: 1.774
- hour: '2023-01-02T07:00:00+01:00'
  price: 2.114
- hour: '2023-01-02T08:00:00+01:00'
  price: 2.164
- hour: '2023-01-02T09:00:00+01:00'
  price: 2.174
- hour: '2023-01-02T10:00:00+01:00'
  price: 2.16
- hour: '2023-01-02T11:00:00+01:00'
  price: 2.139
- hour: '2023-01-02T12:00:00+01:00'
  price: 2.149
- hour: '2023-01-02T13:00:00+01:00'
  price: 2.143
- hour: '2023-01-02T14:00:00+01:00'
  price: 2.184
- hour: '2023-01-02T15:00:00+01:00'
  price: 2.251
- hour: '2023-01-02T16:00:00+01:00'
  price: 2.321
- hour: '2023-01-02T17:00:00+01:00'
  price: 3.663
- hour: '2023-01-02T18:00:00+01:00'
  price: 3.707
- hour: '2023-01-02T19:00:00+01:00'
  price: 3.611
- hour: '2023-01-02T20:00:00+01:00'
  price: 3.505
- hour: '2023-01-02T21:00:00+01:00'
  price: 2.124
- hour: '2023-01-02T22:00:00+01:00'
  price: 2.031
- hour: '2023-01-02T23:00:00+01:00'
  price: 1.904
```

**Comment by woopstar** [comment]
+1 on this one. I will donate coffee when fixed.

**Comment by jonasbkarlsson** [comment]
> +1 on this one. I will donate coffee when fixed.

Thanks for the extra encouragement. That will affect the prioritization.

**Comment by woopstar** [comment]
> Thanks for the extra encouragement. That will affect the prioritization.

You are more than welcome. If it will further affect the prior, I will donate much coffee ;) We are talking LITERS!

**Comment by jonasbkarlsson** [comment]
There is now a pre-release [v1.6.0-dev1](https://github.com/jonasbkarlsson/ev_smart_charging/releases/tag/v1.6.0-dev1) with support of Energi Data Service. It passes all my automatic testing, but I haven't used it for real yet. If someone would like to try it, please let me know how it works!

**Comment by woopstar** [comment]
> If someone would like to try it, please let me know how it works!

Seems config flow might be buggy:

<img width="571" alt="image" src="https://user-images.githubusercontent.com/2997782/213994418-5c9c8f9b-b39d-4fc7-a1a8-85f3c07debc1.png">

**Comment by jonasbkarlsson** [comment]
> Seems config flow might be buggy:

The integration first looks for a Nordpool integration, and then for a Energi Data Service integration. It doesn't generate a list of possible entities, it just suggests the first one it finds. 

You can manually enter the  Energi Data Service sensor entity.

**Comment by woopstar** [comment]
> The integration first looks for a Nordpool integration, and then for a Energi Data Service integration. It doesn't generate a list of possible entities, it just suggests the first one it finds.
> 
> You can manually enter the Energi Data Service sensor entity.

Yes, but the config prompt does not work. It does not show what entity to put int.

I tried to change the sensor in the config file from nordpool to EDS directly and restarted HA. It caused the same error with  Price sensor invalid from here https://github.com/jonasbkarlsson/ev_smart_charging/blob/main/custom_components/ev_smart_charging/coordinator.py#L486

The config used is this:

```
{
   "entry_id":"d1fc4ae3a861907a2dca0ccfa11fc5de",
   "version":2,
   "domain":"ev_smart_charging",
   "title":"EV Smart Charging",
   "data":{
      "price_sensor":"sensor.energi_data_service_auto_tarif",
      "ev_soc_sensor":"sensor.audi_e_tron_state_of_charge",
      "ev_target_soc_sensor":"input_number.audi_e_tron_charging_target",
      "charger_entity":"",
      "pct_per_hour":9.435,
      "start_hour":"21:00",
      "ready_hour":"08:00",
      "maximum_price":0.0,
      "min_soc":15,
      "device_name":"EV Smart Charging"
   },
   "options":{
      "pct_per_hour":9.435,
      "start_hour":"21:00",
      "ready_hour":"08:00",
      "maximum_price":0.0,
      "min_soc":15
   },
   "pref_disable_new_entities":false,
   "pref_disable_polling":false,
   "source":"user",
   "unique_id":null,
   "disabled_by":null
}
```

**Comment by cnrd** [comment]
Running 1.6.0-dev1, I do not have the same experience as @woopstar

All information in the config prompt shows up correctly.
![Screenshot 2023-01-23 at 09 50 20](https://user-images.githubusercontent.com/5996053/213998581-cf2b5d1d-413b-4045-b720-1a0c222d0c13.png)

**Comment by woopstar** [comment]
> Running 1.6.0-dev1, I do not have the same experience as @woopstar
> 
> All information in the config prompt shows up correctly. ![Screenshot 2023-01-23 at 09 50 20](https://user-images.githubusercontent.com/5996053/213998581-cf2b5d1d-413b-4045-b720-1a0c222d0c13.png)

Nice. How did you install 1.6.0-dev1 ? Maybe that's whats causing it

**Comment by cnrd** [comment]
HACS, search for the integration, click redownload and choose the beta version.

**Comment by cnrd** [comment]
It also seems like using energi_data_service is working fine, I have however only tested it by manually changing the sensor through core.config_entries.

**Comment by cnrd** [comment]
> > If someone would like to try it, please let me know how it works!
> 
> Seems config flow might be buggy:
> 
> <img alt="image" width="571" src="https://user-images.githubusercontent.com/2997782/213994418-5c9c8f9b-b39d-4fc7-a1a8-85f3c07debc1.png">

Be aware that it looks like your address is visible in the screenshot (On the Unifi integration)

**Comment by woopstar** [comment]
> > > If someone would like to try it, please let me know how it works!
> > 
> > 
> > Seems config flow might be buggy:
> > <img alt="image" width="571" src="https://user-images.githubusercontent.com/2997782/213994418-5c9c8f9b-b39d-4fc7-a1a8-85f3c07debc1.png">
> 
> Be aware that it looks like your address is visible in the screenshot (On the Unifi integration)

Thanks. Finding my address is not an issue, so did not remove it. But thanks.

**Comment by woopstar** [comment]
I can confirm I have 1.6.0-dev1 working. Found the issue locally.

**Comment by woopstar** [comment]
@jonasbkarlsson would this need to check the platform used? https://github.com/jonasbkarlsson/ev_smart_charging/blob/a1fb786cef9336b2b0492d18ab218ab2e8f0c505/custom_components/ev_smart_charging/coordinator.py#L464

**Comment by jonasbkarlsson** [comment]
> @jonasbkarlsson would this need to check the platform used?
> 
> https://github.com/jonasbkarlsson/ev_smart_charging/blob/a1fb786cef9336b2b0492d18ab218ab2e8f0c505/custom_components/ev_smart_charging/coordinator.py#L464

No, `self.tomorrow_valid` is an internal variable that is set after the price information has been converted to the internal format. It has no dependency to the tomorrow_valid attribute in the price integrations.

**Comment by woopstar** [comment]
> > @jonasbkarlsson would this need to check the platform used?
> > https://github.com/jonasbkarlsson/ev_smart_charging/blob/a1fb786cef9336b2b0492d18ab218ab2e8f0c505/custom_components/ev_smart_charging/coordinator.py#L464
> 
> No, `self.tomorrow_valid` is an internal variable that is set after the price information has been converted to the internal format. It has no dependency to the tomorrow_valid attribute in the price integrations.

The tomorrow is valid is also present in Energi Data Service. I'm pretty sure the Nordpool bug fix is also applied to EDS. Anyways, if you say it's okay, then I assume it is :)

**Comment by woopstar** [comment]
I can confirm that it so far works flawless with EDS:

<img width="510" alt="image" src="https://user-images.githubusercontent.com/2997782/214055060-f163ad97-4fb6-4c77-9994-97fdc7a46977.png">

**Comment by woopstar** [comment]
This implementation now leads to a new Feature Request.

Being able to use forecast data in EDS.

Currently when I calculate when to charge, I use the forecast data if clock is between 00 and 13 since next days prices are not available. So I add the 24 next hours from the forecast data to check if it's cheaper to charge at night.

The list is simply compiled this way:

```
{%- set prices = state_attr("sensor.energi_data_service", "today") + ( state_attr("sensor.energi_data_service", "tomorrow") if now().hour >= 13 else state_attr("sensor.energi_data_service", "forecast")[0:24] | map(attribute='price') | list ) %}
```

**Comment by jonasbkarlsson** [comment]
I will add the EDS forecast data to my todo-list, but there are few other things I would like to do first.

**Comment by jonasbkarlsson** [comment]
Feature Energi Data Service added in release [v1.6.0](https://github.com/jonasbkarlsson/ev_smart_charging/releases/tag/v1.6.0).

**Comment by woopstar** [comment]
Is it me that "thinks" wrong of it, but I see that Apexchart is not updating when the charging schedule has passed? See screenshot:

<img width="504" alt="image" src="https://user-images.githubusercontent.com/2997782/214539190-b7246f64-67ca-4bb3-8be6-ff900edac311.png">

**Comment by jonasbkarlsson** [comment]
Was that charging schedule created yesterday, or did you by any chance reboot HA during that time (06:00-07:00?)?

**Comment by woopstar** [comment]
It was yesterday. I did not reboot.

**Comment by jonasbkarlsson** [comment]
In the orignal implementation, the old schedule was removed when tomorrows prices became available, but I think I have changed that. Or at least I thought so... will need to check.

**Comment by woopstar** [comment]
As promised :)

<img width="583" alt="image" src="https://user-images.githubusercontent.com/2997782/214602113-08d951ba-5bc9-480c-bb53-e0a621f4fceb.png">

**Comment by woopstar** [comment]
> In the orignal implementation, the old schedule was removed when tomorrows prices became available, but I think I have changed that. Or at least I thought so... will need to check.

I have no idea, just glanced through fastly, but could it be that this else section needs to remove the schedule also?
https://github.com/jonasbkarlsson/ev_smart_charging/blob/main/custom_components/ev_smart_charging/coordinator.py#L232

**Comment by jonasbkarlsson** [comment]
This was fixed by #127 and is now part of [v1.7.0-dev1](https://github.com/jonasbkarlsson/ev_smart_charging/releases/tag/v1.7.0-dev1).
